### ISSUE TITLE
Performance impact of Pooling

### ISSUE THREAD

**Issue by prateek** [couchbase/vellum#19]
Hey! A little quick background: I'm using Vellum in a [timeseries database](https://github.com/m3db/m3db) to reverse index timeseries by tags. We use a combination of map backed segments and those using FSTs (using Vellum), periodically compacting map backed segments to FSTs.

I was doing some performance work, because we were seeing about a 20% spike during the period of FST building v steady state. I noticed lots of alloc/inuse space in `func (p *builderNodePool) alloc() *builderNode ` and `func (p *transitionPool) alloc() *transition`. So I reverted https://github.com/couchbase/vellum/commit/5083a469fcef50a4705e3c90652246ecd6806f2a and https://github.com/couchbase/vellum/commit/5c7eccda103b78ded6f8df2d46f4bce903b6a847 and it helped quite a bit. The delta spikes we saw were now around ~8-10%. 

Couple of questions:
- What was the motivation for those commits? Was their a workload they helped a lot?
- Any other advise about tuning Vellum to have more bounded memory usage?

**Comment by prateek** [comment]
@mschoch @steveyen I'd love your input when you have a moment.

**Comment by mschoch** [comment]
Thanks for reporting this.  I'll defer to @steveyen for more detailed explanation, but these were changes that actually improve performance for us.  So it would be very interesting to understand how you're using it, and why it hurt performance for you.

**Comment by prateek** [comment]
@mschoch that's definitely interesting. Would you be up for a 30m video-call sometime next/following week?

**Comment by prateek** [comment]
Re-produced the issue with an isolated file using the `vellum/cmd` tool.




- Time to generate a set using https://github.com/couchbase/vellum/commit/eb6ae3743b3f300f2136f83ca78c08cc071edbd4 (i.e. `master` as of today): **1m15s, max heap in use: 8GB**

```
❯ time ./upstream-vellum --expvar ':8081' --cpuprofile upstream-cpu.out --memprofile upstream-mem.out set --sorted sample.input output.fst
./upstream-vellum --expvar ':8081' --cpuprofile upstream-cpu.out --memprofile  75.95s user 16.53s system 123% cpu 1:14.75 total
```

![screen shot 2018-05-20 at 14 51 31](https://user-images.githubusercontent.com/199982/40282444-6706091a-5c3d-11e8-9303-a7661bd02519.png)

- Time to generate a set using my fork (i.e. https://github.com/couchbase/vellum/commit/eb6ae3743b3f300f2136f83ca78c08cc071edbd4 without https://github.com/couchbase/vellum/commit/5083a469fcef50a4705e3c90652246ecd6806f2a and https://github.com/couchbase/vellum/commit/5c7eccda103b78ded6f8df2d46f4bce903b6a847 ): **1m05s, max heap in use: 12MB**

```
❯ time ./fork-vellum --expvar ':8081' --cpuprofile fork-cpu.out --memprofile fork-mem.out set --sorted sample.input output.fst
./fork-vellum --expvar ':8081' --cpuprofile fork-cpu.out --memprofile  set     81.91s user 2.30s system 131% cpu 1:04.12 total
```

![screen shot 2018-05-20 at 14 55 20](https://user-images.githubusercontent.com/199982/40282518-a00975b6-5c3e-11e8-9f98-e51fc699a27e.png)

The profiles contain symbols (go1.10.2) so you can view them using pprof/go tool without the binary: [profiles.zip](https://github.com/couchbase/vellum/files/2020598/profiles.zip)



<details>
<summary>Statistics about the input file</summary>

```
❯ ls -ltrh sample.input
-rw-r--r--  1 user  group   230M May 20 14:02 sample.input

❯ wc sample.input
 1372640 1372877 241480857 sample.input

# line length distribution
❯ cat sample.input| awk '{print length($0)}' | sort | uniq -c | histogram.py -a
# NumSamples = 1372640; Min = 26.00; Max = 2441.00
# Mean = 100.166683; Variance = 14084.789705; SD = 118.679357; Median 303.500000
# each ∎ represents a count of 15162
    26.0000 -  267.5000 [1137204]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   267.5000 -   509.0000 [235062]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
   509.0000 -   750.5000 [   369]:
   750.5000 -   992.0000 [     1]:
   992.0000 -  1233.5000 [     1]:
  1233.5000 -  1475.0000 [     0]:
  1475.0000 -  1716.5000 [     0]:
  1716.5000 -  1958.0000 [     2]:
  1958.0000 -  2199.5000 [     0]:
  2199.5000 -  2441.0000 [     1]:
```

</details>

**Comment by prateek** [comment]
Spent some more time looking at the code. I better understand what you're trying achieve with the pooling (please correct if I'm wrong): the pooling is useful when you want to re-use the builder to construct multiple FSTs because you avoid a lot repeated allocations. 

In my case however, the pooling is detrimental to performance because the pool grows to a point where it causes more GC issues than the repeated allocations do. 

Is there a logical place where a builderNode can be returned to the pool whilst construct a single FST?

**Comment by mschoch** [comment]
Normally I would, but I'm traveling this week.  Hopefully @steveyen can take a look, you've provided a lot of good detail.

**Comment by steveyen** [comment]
hi @prateek 

(btw, https://github.com/m3db/m3db looks pretty cool!)

> Spent some more time looking at the code. I better understand what you're trying achieve with the pooling (please correct me if I'm wrong): the pooling is useful when you want to re-use the builder to construct multiple FSTs because you avoid a lot repeated allocations.

Indeed -- that was the intention.  We're using vellum in a long running indexing system that might have to chew through lots of "big data" (see https://github.com/blevesearch/bleve), where there's lots of documents made up of lots of terms, and there'd be a vellum FST per term.  Without the builderPool/transitionPool's, the additional memory allocations on every single item would eventually lead to much higher GC activities and impact throughput.

In contrast, I think your benchmarks appear to be relatively quick or short... in the timeframes of a minute or so, there might not be enough activity for GC to kick in.  So, maybe there's a throughput versus latency tradeoff angle that's happening here.

**Comment by prateek** [comment]
Hey @steveyen, thanks for the response!

Pooling helping with the workload you describe makes sense. The benchmark I posted above was to repro the issue locally. Within M3DB, we're using Vellum pretty similarly to how you are using it in Bleve. In steady state - M3DB indexes metrics in Segment's backed by go's `map[...]` as soon as it receives a write, and then periodically converts these maps to FSTs using Vellum. One FST per unique metric tag. 

In my testing the pooling hurt performance in M3DB too. It required a lot more heap (as you'd expect), and created FSTs slower, along with all the other GC issues the extra heap caused. As a result we started splitting our corpus into chunks to make the FSTs being constructed be smaller. Even in this mode, the version without pooling does better for us.

Some questions for you:
- I don't understand all the nuances of the construction code, is there a way to return items to the pools (transitions/builderNodes) within the construction of a FST? 
- What kind of performance impact did you see with pooling enabled? Any memory/latency issues as a result?
- Would you be able to share any stats about your usage? The size of FSTs you construct (number of terms, term length etc), how frequently you compact, etc. I mainly want this to understand if we're running Vellum in a manner that's very different from how you're using it.

**Comment by prateek** [comment]
\*bump\* @steveyen whenever you get a chance

**Comment by prateek** [comment]
@steveyen @mschoch I'd love to followup about this. Would you have some cycles for a quick chat anytime over the upcoming weeks?

**Comment by prateek** [comment]
addressed by #24
